beacon/lib/beacon/scope.ex

```diff
@@ -0,0 +1,200 @@
+defmodule Beacon.Scope do
+  @moduledoc false
+
+  use GenServer
+  require Logger
+
+  @default_broadcast_interval 5_000
+
+  @spec member_counts(atom) :: %{Beacon.group() => non_neg_integer}
+  def member_counts(scope) do
+    scope
+    |> table_name()
+    |> :ets.select([{{:_, :"$1"}, [], [:"$1"]}])
+    |> Enum.reduce(%{}, fn member_counts, acc ->
+      Map.merge(acc, member_counts, fn _k, v1, v2 -> v1 + v2 end)
+    end)
+  end
+
+  @spec member_count(atom, Beacon.group()) :: non_neg_integer
+  def member_count(scope, group) do
+    scope
+    |> table_name()
+    |> :ets.select([{{:_, :"$1"}, [], [:"$1"]}])
+    |> Enum.sum_by(fn member_counts -> Map.get(member_counts, group, 0) end)
+  end
+
+  @spec member_count(atom, Beacon.group(), node) :: non_neg_integer
+  def member_count(scope, group, node) do
+    case :ets.lookup(table_name(scope), node) do
+      [{^node, member_counts}] -> Map.get(member_counts, group, 0)
+      [] -> 0
+    end
+  end
+
+  @spec groups(atom) :: MapSet.t(Beacon.group())
+  def groups(scope) do
+    scope
+    |> table_name()
+    |> :ets.select([{{:_, :"$1"}, [], [:"$1"]}])
+    |> Enum.reduce(MapSet.new(), fn member_counts, acc ->
+      member_counts
+      |> Map.keys()
+      |> MapSet.new()
+      |> MapSet.union(acc)
+    end)
+  end
+
+  @typep member_counts :: %{Beacon.group() => non_neg_integer}
+
+  defp table_name(scope), do: :"#{scope}_beacon_peer_counts"
+
+  defmodule State do
+    @moduledoc false
+    @type t :: %__MODULE__{
+            scope: atom,
+            message_module: module,
+            broadcast_interval: non_neg_integer,
+            peer_counts_table: :ets.tid(),
+            peers: %{pid => reference}
+          }
+    defstruct [
+      :scope,
+      :message_module,
+      :broadcast_interval,
+      :peer_counts_table,
+      peers: %{}
+    ]
+  end
+
+  @spec start_link(atom, Keyword.t()) :: GenServer.on_start()
+  def start_link(scope, opts \\ []), do: GenServer.start_link(__MODULE__, [scope, opts])
+
+  @impl true
+  def init([scope, opts]) do
+    :ok = :net_kernel.monitor_nodes(true)
```

Comment:
Keep in mind there is an annoying pitfall with monitor nodes, in that the node may be up, but not the Beacon service running on the node! So my suggestion is to not rely on `nodeup` events and instead, when the `Beacon.Scope` process is ready, it broadcasts to everyone.